client/comms/wsconn_test.go

```diff
@@ -0,0 +1,332 @@
+	log.Infof("Generating TLS certificates...")
+
+	org := "dcrdex autogenerated cert"
+	validUntil := time.Now().Add(10 * 365 * 24 * time.Hour)
+	cert, key, err := certgen.NewTLSCertPair(elliptic.P521(), org,
+		validUntil, altDNSNames)
+	if err != nil {
+		return err
+	}
+
+	// Write cert and key files.
+	if err = ioutil.WriteFile(certFile, cert, 0644); err != nil {
+		return err
+	}
+	if err = ioutil.WriteFile(keyFile, key, 0600); err != nil {
+		os.Remove(certFile)
+		return err
+	}
+
+	log.Infof("Done generating TLS certificates")
+	return nil
+}
+
+func TestWsConn(t *testing.T) {
+	upgrader := websocket.Upgrader{}
+	pingCh := make(chan struct{})
+	readPumpCh := make(chan interface{})
+	writePumpCh := make(chan *msgjson.Message)
+	ctx, cancel := context.WithCancel(context.Background())
+
+	handler := func(w http.ResponseWriter, r *http.Request) {
+		c, err := upgrader.Upgrade(w, r, nil)
+		if err != nil {
+			t.Fatalf("unable to upgrade http connection: %s", err)
+		}
+
+		defer c.Close()
+
+		go func() {
+			for {
+				select {
+				case <-pingCh:
+					err = c.WriteControl(websocket.PingMessage, []byte{},
+						time.Time{})
+					if err != nil {
+						fmt.Printf("unable to write ping: %v", err)
+						return
+					}
+
+				case msg := <-readPumpCh:
+					err := c.WriteJSON(msg)
+					if err != nil {
+						fmt.Printf("unable to write message: %v", err)
+						return
+					}
+
+				case <-ctx.Done():
+					c.Close()
+					return
+				}
+			}
+		}()
+
+		for {
+			mType, message, err := c.ReadMessage()
+			if err != nil {
+				if opErr, ok := err.(*net.OpError); ok {
+					if opErr.Op == "read" {
+						return
+					}
+				}
+
+				if websocket.IsCloseError(err) {
+					return
+				}
+
+				if nErr, ok := err.(net.Error); ok {
+					if nErr.Timeout() {
+						return
+					}
+				}
+
+				log.Errorf("unable to read message: %v", err)
+				return
+			}
+
+			if mType == websocket.TextMessage {
+				msg, err := msgjson.DecodeMessage(message)
+				if err != nil {
+					fmt.Printf("unable to decode msg: %v", err)
+					return
+				}
+
+				writePumpCh <- msg
+			}
+		}
+
+	}
+
+	certFile, err := ioutil.TempFile("", "certfile")
+	if err != nil {
+		t.Fatalf("unable to create temp certfile: %s", err)
+	}
+	certFile.Close()
+	defer os.Remove(certFile.Name())
+
+	keyFile, err := ioutil.TempFile("", "keyfile")
+	if err != nil {
+		t.Fatalf("unable to create temp keyfile: %s", err)
+	}
+	keyFile.Close()
+	defer os.Remove(keyFile.Name())
+
+	err = genCertPair(certFile.Name(), keyFile.Name(), nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	host := "127.0.0.1:6060"
+	http.HandleFunc("/ws", handler)
+
+	go func() {
+		err = http.ListenAndServeTLS(host, certFile.Name(), keyFile.Name(), nil)
+		if err != nil {
+			fmt.Println(err)
+		}
+	}()
+
+	cfg := &WsCfg{
+		Host:               host,
+		Path:               "ws",
+		PingWait:           3 * time.Second,
+		WriteWait:          3 * time.Second,
```

Comment:
Can likely go to milliseconds or 10s of milliseconds here so that your forced timeout close below can be faster.